python/ray/serve/_private/result_wrapper.py

```diff
@@ -0,0 +1,126 @@
+import threading
+import time
+from abc import ABC, abstractmethod
+from typing import Callable, Optional
+
+import ray
+from ray.serve._private.utils import calculate_remaining_timeout
+
+
+class ResultWrapper(ABC):
+    @abstractmethod
+    def get(self, timeout_s: Optional[float]):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def get_async(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def next(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def anext(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def resolve_generator_to_ref_sync(self, timeout_s: Optional[float]):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def resolve_generator_to_ref_async(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def resolve_as_top_level_arg(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def add_callback(self, callback: Callable):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def cancel(self):
+        raise NotImplementedError
+
+
+class ActorResultWrapper(ResultWrapper):
+    def __init__(self, obj_ref_or_gen, is_gen: bool):
+        self._obj_ref = None
+        self._obj_ref_gen = None
+        self._is_gen = is_gen
+        self._object_ref_or_gen_sync_lock = threading.Lock()
+
+        if isinstance(obj_ref_or_gen, ray.ObjectRefGenerator):
```

Comment:
I feel it's not an important distinction between saving the object to `self._obj_ref_gen` or `self._obj_ref`. The caller would still need to know whether the wrapper is created with a generator or not and calling into the related methods (`get()` vs `next()`...etc)

Reply:
IMO the code is kind of confusing when you're always using a variable named `obj_ref_or_gen`, because it's not clear in each function whether you expect it to be a normal `ray.ObjectRef` or `ray.ObjectRefGenerator`. Especially now that this result wrapper is abstracted to only deal directly with ray objects, I don't think this adds noticeable complexity and make the code easier to read.

Reply:
If that's the case, can we add some assertions to methods like `get()` and `next()`...etc to explicitly raise those can only be called when initialized with the corresponding objects?

Reply:
yeah thats a good idea, added